lib/coach/middleware_validator.rb

```diff
@@ -21,14 +22,18 @@ def validated_provides!
 
     private
 
+    attr_reader :previous_middlewares
+
     def missing_requirements
-      @middleware.requirements - provided_by_chain
+      @missing_requirements ||= @middleware.requirements - provided_by_chain
     end
 
     def provided_by_chain
       @provided_by_chain ||=
         middleware_dependencies.reduce(@already_provided) do |provided, middleware|
-          provided + self.class.new(middleware, provided).validated_provides!
+          validator = self.class.new(middleware, previous_middlewares, provided)
+          previous_middlewares << middleware
```

Comment:
That's way better. Wasn't too happy with the side effects either, this is much nicer!